### ISSUE TITLE
Base64 encode

### ISSUE THREAD

**Issue by ythy** [ythy/blog#364]
[reference](https://base64.guru/learn/base64-algorithm/encode)
The Base64 encode algorithm converts any data into plain text. Technically, it can be said that it converts eight-bit bytes into six-bit bytes. 
For example, you have the `“我S”` string and want to convert it to Base64:
1. First, you need to split the string letter by letter. Thus, you got 2 groups:
```
我
S
```
2. Next you need to convert each group to binary. To do this, for each letter you need to find the corresponding binary value in the ASCII table. Thus, now you have 2 groups of ones and zeros:
```
111001101000100010010001 (note that letter “我” is a 3-byte character)
01010011
```
3. Now concatenate all the binary values together (that is, glue all the groups along and make sure you get a total of 32 characters):
```
11100110100010001001000101010011
```
4. Then, divide the resulting string into groups so that each one has 6 characters (if the last group has less than 6 characters, you need to fill it with zeros until it reaches the desired length). Well and good, now you have 6 groups:
```
111001
101000
100010
010001
010100
110000 ← append 4 zeros to normalize the six-bit byte
```
5. At this step you have to convert six-bit bytes into eight-bit bytes. To do this, prepend the prefix “00” (two zeros) in front of each group:
```
00111001
00101000
00100010
00010001
00010100
00110000
```
6. There you have to convert each group from binary to decimal by finding its corresponding decimal value in the ASCII table. If you did everything right, each group will be transformed into its integer number as follows:
```
57
40
34
17
20
48
```
7. Integer numbers obtained in the previous step are called “Base64 indices”. They are easy to remember, because it is a zero-based numbering, where each index corresponds to a Latin letter. It starts with the letter “A” in alphabetical order (i.e., A=0, B=1, C=2, D=3, and so on). For complete list, see Base64 Characters Table. So, matching indexes, convert them to corresponding letters:
```
5
o
i
R
U
w
```
8. The final chord, concatenate all letters to get the Base64 string:
```
5oiRUw
```

To summarize, you learned that encoding “我S” to Base64 yields the result “5oiRUw”.
